### ISSUE TITLE
[4.x]: Asset previews doesn't work for filesystems without URLs

### ISSUE THREAD

**Issue by mmikkel** [craftcms/cms#11686]
### Description

When attempting to preview an image (either via the "Preview" element index action or via the Preview button on top of edit page thumbnails) belonging to a filesystem without public URLs, the following error is thrown (displayed as a JS alert dialog):

```
craft\services\Assets::getImagePreviewUrl(): Return value must be of type string, null returned
```

### Steps to reproduce

1. Create a volume, and attach a filesystem with the "Files in this filesystem have public URLs" setting disabled to it
2. Upload an image to the volume
3. Preview the image, either via the asset element index "Preview" action or the "Preview" button on top of the image thumbnail in the asset's edit page

### Expected behavior

The preview modal should appear.

### Actual behavior

No preview modal appears; a JS error dialog appears.

### Craft CMS version

4.1.4.1

### PHP version

8.1.8

### Operating system and version

macOS 12.4, Edge 103.0.1264.62 (Official build) (arm64)

### Database type and version

MariaDB 10.8.3 

### Image driver and version

Imagick 3.7.0 (ImageMagick 7.1.0-29)

### Installed plugins and versions

None

**Comment by hunwalk** [comment]
This is way hacky (please do not edit into your vendor dir like that), but the solution could be base64 encoding images. Resulting in serious DOM sizes but it works

Here: 
https://github.com/craftcms/cms/blob/2b2fc8cd521dfd2238e5e50c355f39cd633c6b3f/src/elements/Asset.php#L1523-L1525

Modify it like this: 

```php
if ($fsNoUrls || $noFolder || $transformNoUrl) {
   return 'data:image/jpg;base64,'.base64_encode(file_get_contents(Craft::getAlias($volume->getFs()->path.'/'.$this->getPath())));
}
```

P.S.: Why isn't there a controller action that would check against every asset url. We also know is something `isCpRequest`, so it would not be a problem. If a filesystem has no urls, getUrl should return a specific url like, domain.tld/admin/asset/fetch?assetUid=<assetUid>

**Comment by mmikkel** [comment]
> P.S.: Why isn't there a controller action that would check against every asset url. We also know is something isCpRequest, so it would not be a problem. If a filesystem has no urls, getUrl should return a specific url like, domain.tld/admin/asset/fetch?assetUid=

Yeah, that's the way it worked in Craft 3 – probably just an oversight in Craft 4 as there has a lot of refactoring for assets under the hood.

**Comment by mmikkel** [comment]
Looked into this a bit more. Seems like the `AssetsController::actionThumb()` action (which is used to generate previews and thumbnails for assets w/o URLs on Craft 3) is removed in Craft 4. 

Basically, a preview or thumbnail request will end up deferring to the volume's transform filesystem - and if that filesystem doesn't have URLs enabled, it will either return an icon URL ([for thumbnails](https://github.com/craftcms/cms/blob/develop/src/services/Assets.php#L627)) or fail w/ the above exception (for image previews and the image editor).

**Comment by hunwalk** [comment]
For the time being here is a plug&play hotfix until you guys resolve the problem
https://github.com/hunwalk/craft-asset-preview-fix

**Comment by brandonkelly** [comment]
This change was semi-intentional… the goal was to centralize all image manipulation operations around image transforms in Craft 4, so extending Craft’s transform functionality would be a little more straightforward. The plan is to get it to a point where we can invoke the transformer even if the filesystem doesn’t have public URLs, and serve the transformed file contents directly from PHP if that’s the only option. We’re not quite there yet though.

So for now I’ve just resolved this by reading out the full image contents, which might end up being heavy depending on the asset, but still better than nothing.

That fix will be in Craft 4.2.1, which comes on Tuesday. If you want to test it early, you can change your `craftcms/cms` requirement in composer.json to `dev-develop as 4.2.0.2` and run `composer update`.

**Comment by brandonkelly** [comment]
Craft 4.2.1 is out with that fix.
